src/core/server/elasticsearch/elasticsearch_service.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ConnectableObservable, Observable, Subscription } from 'rxjs';
+import { filter, map, publishReplay, switchMap } from 'rxjs/operators';
+import { CoreContext, CoreService } from '../../types';
+import { Logger } from '../logging';
+import { ClusterClient } from './cluster_client';
+import { ElasticsearchClientConfig } from './elasticsearch_client_config';
+import { ElasticsearchConfig } from './elasticsearch_config';
+
+interface CoreClusterClients {
+  config: ElasticsearchConfig;
+  adminClient: ClusterClient;
+  dataClient: ClusterClient;
+}
+
+/** @internal */
+export interface ElasticsearchServiceStartContract {
+  // Required for the BWC only.
+  readonly bwc: {
```

Comment:
Sure, here are the places where bwc config is used right now:

### ES plugin:
- requestTimeout
- shardTimeout
- apiVersion
- healthcheck.delay
- [_using entire config as default in `createCluster` when not provided_](https://github.com/elastic/kibana/pull/28344/files#diff-d6a295c68fe4899310d2a439553cdbfaR75)

### Timelion plugin:

- shardTimeout

### Console plugin:

- requestTimeout
- hosts
- requestHeadersWhitelist
- customHeaders
- ssl (all properties)

Once we move client part of legacy es plugin to the core and somehow expose ES proxy route from the core itself so that Console can just use it, we'll need to take care of just two use cases before we can remove `bwc`: 

* `shardTimeout` for Timelion - maybe expose this from contract?
* default config for legacy `createCluster` - we can do that inside of the core instead, but in core config is `Observable`, so core's `createClient` will return `Observable` as well, while legacy `createCluster` should synchronously return `Cluster`. The first solution that comes into my mind is using Observable `clusterClient$` inside of legacy `Cluster` and since all of it's methods are async, we can retrieve client under the hood when it's needed for the first time.

What do you think?

P.S.: Initially I had `requestTimeout`, `shardTimeout` and `apiVersion` in the main ES Service start contract, but later changed my mind and removed them as I wasn't sure whether it can have any value for consumer. Theoretically we can expose `config$: Observalbe<subset of config that can be useful for consumers>` if/when we need it.

Reply:
Makes sense. Maybe `legacy` would be a better name than `bwc` for the sake of consistency. What do you think?

Reply:
> Maybe `legacy` would be a better name than `bwc` for the sake of consistency.
Yep, agree, will change.